crates/pathfinder/src/consensus/inner/p2p_task_tests.rs

```diff
@@ -0,0 +1,1402 @@
+        ProposalPart::ProposalCommitment(p2p_proto::consensus::ProposalCommitment {
+            block_number: height,
+            parent_commitment: zero_hash,
+            builder: zero_address,
+            timestamp: 1000,
+            protocol_version: "0.0.0".to_string(),
+            old_state_root: zero_hash,
+            version_constant_commitment: zero_hash,
+            state_diff_commitment: p2p_proto::common::Hash(proposal_commitment.0), /* Use real commitment */
+            transaction_commitment: zero_hash,
+            event_commitment: zero_hash,
+            receipt_commitment: zero_hash,
+            concatenated_counts: Felt::ZERO,
+            l1_gas_price_fri: 0,
+            l1_data_gas_price_fri: 0,
+            l2_gas_price_fri: 0,
+            l2_gas_used: 0,
+            next_l2_gas_price_fri: 0,
+            l1_da_mode: p2p_proto::common::L1DataAvailabilityMode::Calldata,
+        })
+    }
+
+    /// ProposalFin deferred until TransactionsFin is processed.
+    ///
+    /// **Scenario**: ProposalFin arrives before TransactionsFin. Execution has
+    /// started (TransactionBatch received), so ProposalFin must be deferred
+    /// until TransactionsFin arrives and is processed, then finalization
+    /// can proceed.
+    ///
+    /// **Test**: Send Init → BlockInfo → TransactionBatch →
+    /// ProposalCommitment → ProposalFin → TransactionsFin.
+    ///
+    /// Verify ProposalFin is deferred (no proposal event), then verify
+    /// finalization occurs after TransactionsFin arrives. Also verify
+    /// ProposalFin is persisted in the database even when deferred.
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_proposal_fin_deferred_until_transactions_fin_processed() {
+        let chain_id = ChainId::SEPOLIA_TESTNET;
+        let validator_address = ContractAddress::new_or_panic(Felt::from_hex_str("0x123").unwrap());
+        let mut env = TestEnvironment::new(chain_id, validator_address);
+        env.create_committed_parent_block(1);
+        env.wait_for_task_initialization().await;
+
+        let proposer_address = ContractAddress::new_or_panic(Felt::from_hex_str("0x456").unwrap());
+        let height_and_round = HeightAndRound::new(2, 1);
+        let transactions = create_transaction_batch(0, 5, chain_id);
+        let (proposal_init, block_info) =
+            create_test_proposal(chain_id, 2, 1, proposer_address, transactions.clone());
+
+        // Focus is on batch execution and deferral logic, not commitment validation.
+        // Using a dummy commitment...
+        let proposal_commitment = ProposalCommitment(Felt::ZERO);
+
+        // Step 1: Send ProposalInit
+        env.p2p_tx
+            .send(Event::Proposal(
+                height_and_round,
+                ProposalPart::Init(proposal_init),
+            ))
+            .expect("Failed to send ProposalInit");
+        tokio::time::sleep(Duration::from_millis(300)).await;
```

Comment:
What are these sleeps for...? They add up to almost 2s of additional test execution..

Reply:
Probably just me copying and pasting after I got the first one working. Will review and see if we can nuke them or substantially reduce them. Good catch!